cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/ProgramLifecycleHttpHandlerTest.java

```diff
@@ -309,6 +312,38 @@ public void testMapreduceHistory() throws Exception {
   }
 
   /**
+   * Tests history of a non existing program
+   */
+  @Test
+  public void testNonExistingProgramHistory() throws Exception {
+
+    Id.Program program = Id.Program.from(TEST_NAMESPACE2, DUMMY_APP_ID, ProgramType.MAPREDUCE, DUMMY_MR_NAME);
+    String namespace = program.getNamespaceId();
+    try {
+      deploy(DummyAppWithTrackingTable.class, Constants.Gateway.API_VERSION_3_TOKEN, namespace);
+      // first run
+      startProgram(program);
+      waitState(program, ProgramRunStatus.RUNNING.toString());
+      final String statusUrl = getVersionedAPIPath("apps/" + program.getApplicationId() +
```

Comment:
I don't think you need to add a new test method for this which deploys app and stuff. All you need to test is that a non existing app runs method returns NotFound status.

You can just add this (327 to 331) in another appropriate test where we are checking history of some program. For example testWorkflowHistory. Even if you decide to have a  separate test for this you dont need to start the program and wait for it complete. You can just deploy the app and check with non existing program for NotFound status code. Running the mapreduce program will slow down the build by its runtime.